backend/msgpack/server.go

```diff
@@ -0,0 +1,326 @@
+) error {
+	mu := unaggregated.MetricUnion{
+		Type:          unaggregated.BatchTimerType,
+		ID:            id,
+		BatchTimerVal: val,
+	}
+	return s.writeMetric(mu, vp)
+}
+
+func (s *server) WriteGaugeWithPolicies(
+	id []byte,
+	val float64,
+	vp policy.VersionedPolicies,
+) error {
+	mu := unaggregated.MetricUnion{
+		Type:     unaggregated.GaugeType,
+		ID:       id,
+		GaugeVal: val,
+	}
+	return s.writeMetric(mu, vp)
+}
+
+func (s *server) Flush() error {
+	s.RLock()
+	if s.state != serverOpen {
+		s.RUnlock()
+		return errServerIsNotOpenOrClosed
+	}
+	multiErr := xerrors.NewMultiError()
+	for shard, encoder := range s.encodersByShard {
+		encoder.Lock()
+		bufferedEncoder := encoder.Encoder()
+		buffer := bufferedEncoder.Buffer()
+		if buffer.Len() == 0 {
+			encoder.Unlock()
+			continue
+		}
+		newBufferedEncoder := s.encoderPool.Get()
+		newBufferedEncoder.Reset()
+		encoder.Reset(newBufferedEncoder)
+		encoder.Unlock()
+		if err := s.writeTo(uint32(shard), bufferedEncoder); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+	s.RUnlock()
+	return multiErr.FinalError()
+}
+
+func (s *server) Close() error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state != serverOpen {
+		return errServerIsNotOpenOrClosed
+	}
+	s.state = serverClosed
+	return s.topology.Close()
+}
+
+func (s *server) writeMetric(mu unaggregated.MetricUnion, vp policy.VersionedPolicies) error {
+	shard, err := s.topology.Shard(mu.ID)
+	if err != nil {
+		return err
+	}
+	encoder, err := s.getOrCreateEncoder(shard)
+	if err != nil {
+		return err
+	}
+	encoder.Lock()
+
+	var (
+		bufferedEncoder = encoder.Encoder()
+		buffer          = bufferedEncoder.Buffer()
+		sizeBefore      = buffer.Len()
+	)
+
+	// NB(xichen): it is a bit unfortunate that we need to convert the concrete
+	// metric type to the union type, pass it in, and convert the union type
+	// back to the concrete metric type due to lack of generics to avoid repeating
+	// the same logic multiple times in the type-specific APIs.
+	switch mu.Type {
+	case unaggregated.CounterType:
+		cp := unaggregated.CounterWithPolicies{
+			Counter:           mu.Counter(),
+			VersionedPolicies: vp,
+		}
+		err = encoder.EncodeCounterWithPolicies(cp)
+	case unaggregated.BatchTimerType:
+		btp := unaggregated.BatchTimerWithPolicies{
+			BatchTimer:        mu.BatchTimer(),
+			VersionedPolicies: vp,
+		}
+		err = encoder.EncodeBatchTimerWithPolicies(btp)
+	case unaggregated.GaugeType:
+		gp := unaggregated.GaugeWithPolicies{
+			Gauge:             mu.Gauge(),
+			VersionedPolicies: vp,
+		}
+		err = encoder.EncodeGaugeWithPolicies(gp)
+	default:
+		err = errUnrecognizedMetricType
+	}
+
+	if err != nil {
+		s.log.WithFields(
+			xlog.NewLogField("metric", mu),
+			xlog.NewLogField("policies", vp.String()),
+			xlog.NewLogErrField(err),
+		).Error("encode metric with policies error")
+		// Rewind buffer and clear out the encoder error.
+		buffer.Truncate(sizeBefore)
+		encoder.Reset(bufferedEncoder)
```

Comment:
The encoder stops encoding when it runs into an error so you need to call reset to clear the internal error in the encoder.